File: January/cpp/Maximum difference between node and ancestor.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    int diff(TreeNode* root,int currmin,int currmax){
        if (root==NULL){
            return currmax-currmin;
        }
        currmax = max(currmax,root->val);
        currmin = min(currmin,root->val);
        int left = diff(root->left, currmin,currmax);
        int right = diff(root->right,currmin,currmax);
        return max(left,right);
    }
public:
    int maxAncestorDiff(TreeNode* root) {
        return diff(root,root->val,root->val);
    }
};

File: January/Java/Maximum Difference Between Node and Ancestor.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int maxAncestorDiff(TreeNode root) {
        return dfs(root, root.val, root.val);
    }

    public int dfs(TreeNode node, int currmax, int currmin) {
        if (node == null) {
            return currmax - currmin;
        }
        currmax = Math.max(currmax, node.val);
        currmin = Math.min(currmin, node.val);
        int left = dfs(node.left, currmax, currmin);
        int right = dfs(node.right, currmax, currmin);
        return Math.max(left, right);
    }
}
